Use the fetch stack for documents.

Ms2ger opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13742)

<!-- Reviewable:end -->

Ms2ger committed:
Use fetch in ScriptThread.

I did not find the prose that defines the creation of the request in the HTML
standard (if it exists).

Ms2ger committed:
Remove AsyncResponseListener.

Ms2ger committed:
Improve navigate-child-function.html and navigate-child-function-src.html.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/script_thread.rs, components/script/dom/document.rs, components/script/dom/servoparser/mod.rs, components/script/document_loader.rs

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit d97ae85 with merge 45ab9d5...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/265)

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 19bd43f with merge c8cdccb...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/242)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13750) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
Status update: investigating the timeout in `/_mozilla/mozilla/referrer-policy/unset-referrer-policy/meta-referrer/same-origin/http-https/iframe-tag/upgrade-protocol.keep-origin-redirect.http.html`. Apparently the load event is not dispatched at the iframe created in `queryIframe`.

Ms2ger commented:
The issue lies within this lovely piece of code:
```rust
        let mut ssl_error = None;
        let metadata = match meta_result {
            Ok(meta) => Some(meta),
            Err(NetworkError::SslValidation(url, reason)) => {
                ssl_error = Some(reason);
                let mut meta = Metadata::default(url);
                let mime: Option<Mime> = "text/html".parse().ok();
                meta.set_content_type(mime.as_ref());
                Some(meta)
            },
            Err(_) => None,
        };
```

With the old stack, we get `Err(NetworkError::SslValidation(..))`, create a new `Metadata` and go on our merry way. With the fetch stack, we get some other `Err(..)` value, `ScriptThread::page_headers_available` notices `metadata` is `None` and returns `None` itself, so we bail out without a parser object, and the load event is never fired.

@jdm, thoughts?

jdm commented:
I propose that we modify the `ResponseType::Error` variant to store a `NetworkError` value, and modify the block in `http_network_fetch` that gets an Err value from `obtain_response` to turn the response into an error response that contains the actual error that was returned from the network stack. This will allow the `metadata` method to return an error that was extracted from the error response, rather than just returning an internal error.

Ms2ger commented:
Rebased; planning to look into your suggestion tomorrow. Thanks!

Ms2ger committed:
Store the NetworkError in network error Responses.

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit addec42 with merge 6a83c01...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3669)

Ms2ger committed:
Store the referrer in the Response and return it in Response::metadata().

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ac3dec9 with merge 1286a2b...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3671)

Ms2ger committed:
Make the status handling in http_fetch more robust.

Ms2ger committed:
Return a network error Response from http_network_fetch when obtain_response fails.

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 53046d2 with merge 5d14791...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/322)

Ms2ger committed:
Pass the entire FetchContext to http_network_fetch.

Ms2ger committed:
Set cookies in http_network_fetch.

Ms2ger committed:
Only pass the headers to set_cookies_from_response.

I will need to call it from the fetch code, which does not have a
HttpResponse on hand.

Also, rename the function to match its new signature.

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 23b2f6f with merge 0d98603...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/324)

Ms2ger commented:
r? @jdm

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13854) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13848) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13891) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented on components/net_traits/response.rs:
```diff
@@ -135,11 +135,18 @@ impl Response {
         }
     }
 
+    pub fn get_network_error(&self) -> Option<&NetworkError> {
```
Every caller clones the error, so I'm not sure it's worth returning a borrow.

Ms2ger replied:
I think it's better to at least make it possible for new callers to not have to clone, if they don't need to.

jdm commented on components/net/fetch/methods.rs:
```diff
@@ -1148,8 +1148,14 @@ fn http_network_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
 
+    // Step 14.
+    if credentials_flag {
+        // TODO: should block cookies?
```
I don't think this is a meaningful TODO given https://tools.ietf.org/html/rfc6265#section-7.

jdm commented on components/script/script_thread.rs:
```diff
@@ -2114,30 +2114,29 @@ impl ScriptThread {
-        }, LoadConsumer::Listener(response_target), None)).unwrap();
+            headers: load_data.headers,
+            body: load_data.data,
+            .. RequestInit::default()
```
Let's file a followup issue to support non-follow redirect modes for RequestInit.

jdm reviewed (changes requested):
Nice work; this was a straightforward series of commits to read through.

Ms2ger commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit c7636a2 has been approved by `jdm`

<!-- @bors-servo r=jdm c7636a215ebb63bc87ee0176949291f96659598b -->

bors-servo commented:
:hourglass: Testing commit c7636a2 with merge ba2fb4e...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3825), [arm64](http://build.servo.org/builders/arm64/builds/3835), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3913), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/553), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/555), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3903), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3895), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/584), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/591), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1817)
